### PR TITLE
feat: generic shuffler over RecordBatchStream 

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

**Comment by eddyxu** [comment]
It is still in `lance` crate because `Shuffler`  depends on `FileWriter` which will be migrated to `lance-core` later. Once the migration is done, we can move `Shuffler` to `lance-index`.

File: rust/lance/src/index/vector/ivf/builder.rs [hunk 1]
```diff
@@ -15,109 +15,109 @@
...
+    Ivf,
+};
 use crate::{io::RecordBatchStream, Error, Result};
 
 use super::RESIDUAL_COLUMN;
 
-/// Filter a batch by a range of partition IDs, specified by `part_range`.
-///
-/// Expect the input batch has schema as
+/// Disk-based shuffle a strema of [RecordBatch] into each IVF partition.
```
Comment (westonpace) [inline rust/lance/src/index/vector/ivf/builder.rs:43]:
```suggestion
/// Disk-based shuffle a stream of [RecordBatch] into each IVF partition.
```

File: rust/lance/src/index/vector/ivf/builder.rs [hunk 2]
```diff
@@ -15,109 +15,109 @@
...
-        .values()
-        .iter()
-        .enumerate()
-        .for_each(|(idx, &part_id)| {
-            if part_range.contains(&part_id) {
-                parted_map.entry(part_id).or_default().push(idx as u32);
-            }
-        });
+    ivf: Arc<lance_index::vector::ivf::Ivf>,
+    // TODO: make this auto-generated from transformer chain.
```
Comment (westonpace) [inline rust/lance/src/index/vector/ivf/builder.rs:60]: Make what auto-generated? `num_sub_vectors`?

**Review by westonpace — APPROVED** [review]

**Comment by westonpace** [comment]
Is the goal here to reduce the memory used while training an index?

**Comment by eddyxu** [comment]
> Is the goal here to reduce the memory used while training an index?

The goal is that it has a general shuffler function for both create index and append index:

For create index:

1. shuffle data
2. write per partition

For appending index:
1. shuffle new data
2. Load each existing partition and the new partition from shuffler
3. write this partition.

File: rust/lance/src/index/vector/ivf/builder.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/index/vector/ivf/builder.rs:60]: The end of the transform chain should produce a schema. Lemme add some comments.

For example,  Using BQ will lead to different schema than PQ.

**Pull request merged**